### PR TITLE
OpenVINO backend

### LINKED ISSUE #1682: Intel OpenVINO backend
Hi! I'm working on [OpenVINO](https://github.com/openvinotoolkit/openvino) backend for DeepLabCut for better efficiency. Currently, I validate integration on https://github.com/DeepLabCut/DeepLabCut/blob/master/examples/testscript_pretrained_models.py. However, I cannot check the accuracy because this sample runs human body estimation model on `reachingvideo1` which is a mouse video. May I ask to share the link to "human dancing" video from http://www.mackenziemathislab.org/dlc-modelzoo or can I run a different network on mouse video?

image

WIP branch: https://github.com/dkurt/DeepLabCut/tree/openvino

### PR DESCRIPTION
**Pull request by dkurt**
This PR proposes Intel OpenVINO library integration for inference. OpenVINO uses tool called Model Optimizer (MO) to convert a frozen `.pb` graph to OpenVINO format.

Added `use_openvino` flag which accepts a name of device for OpenVINO (e.g. "CPU", "GPU" or "MULTI:CPU,GPU"). 

Experimented on `Intel(R) Core(TM) i7-6700K` with iGPU on `human_dancing-teamDLC`:

```
Duration of video [s]:  6.9 , recorded with  28.56 fps!
Overall # of frames:  197  found with (before cropping) frame dimensions:  1280 720
```

Backend | Device | Processing time 
---|---|---
TensorFlow 2.7.1 | CPU | 03m39s (0.89 FPS)
OpenVINO 2022.1 | CPU | 02m26s (1.34 FPS)
OpenVINO 2022.1 | GPU | 03m17s (1.00 FPS)
OpenVINO 2022.1 | MULTI:CPU,GPU | 01m44s (1.89 FPS)

resolves https://github.com/DeepLabCut/DeepLabCut/issues/1682

### CODE REVIEW

**Comment by MMathisLab** [comment]
hi @dkurt thanks! Could I also ask you add some documentation, i.e. how to use, could be under a "recipe" in the cookbook, which just requires you making a .md file https://github.com/DeepLabCut/DeepLabCut/tree/master/docs/recipes and then docstring for the main function

**Comment by dkurt** [comment]
> hi @dkurt thanks! Could I also ask you add some documentation, i.e. how to use, could be under a "recipe" in the cookbook, which just requires you making a .md file https://github.com/DeepLabCut/DeepLabCut/tree/master/docs/recipes and then docstring for the main function

Thanks! Added a guide.

**Comment by dkurt** [comment]
Hi, @MMathisLab! May I ask review it again?

**Comment by dkurt** [comment]
@MMathisLab, may I ask review it again?

**Comment by MMathisLab** [comment]
@jeylau my only question is if we should package like NWB; thus not a core dependency, but if used, the user is asked to install openVINO ? thoughts? 

Thanks @dkurt for the cool contribution

**Comment by dkurt** [comment]
Changed OpenVINO to be an optional dependency so it can be installed by `pip install deeplabcut[openvino]`. Also, updated PR to the latest OV version.

**Comment by jeylau** [comment]
I agree @MMathisLab, and I believe `pip install deeplabcut[openvino]` is a great solution!

File: deeplabcut/pose_estimation_tensorflow/core/openvino/session.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        if not os.path.exists(self.xml_path):
+            subprocess.run(
+                [
+                    "mo",
+                    "--output_dir",
+                    os.path.dirname(cfg["init_weights"]),
+                    "--input_model",
+                    cfg["init_weights"] + ".pb",
+                    "--input_shape",
+                    "[1, 747, 832, 3]",
```
Comment (jeylau) [inline deeplabcut/pose_estimation_tensorflow/core/openvino/session.py:30]: Should image shape be hardcoded?

File: requirements.txt [hunk 2]
```diff
@@ -25,3 +25,4 @@ tf_slim
 tqdm
 moviepy
 Pillow>=7.1
+openvino-dev==2022.1.0
```
Comment (jeylau) [inline requirements.txt]: Is this necessary in the requirements? This is the file that is parsed when GitHub runs CI, maybe we could leave it out. Please tell me if I'm missing something :)

**Review by jeylau — CHANGES_REQUESTED** [review]

File: deeplabcut/pose_estimation_tensorflow/core/openvino/session.py (hunk 1 above)
Comment (dkurt) [reply deeplabcut/pose_estimation_tensorflow/core/openvino/session.py:30]: No, this is just a default value. Later network is reshaped to actual image sizes (see `self.net.reshape`)

Comment (jeylau) [reply deeplabcut/pose_estimation_tensorflow/core/openvino/session.py:30]: Got it!

File: requirements.txt (hunk 2 above)
Comment (dkurt) [reply requirements.txt]: Sure, it can be removed

**Review by jeylau — APPROVED** [review]

**Review by AlexEMG — APPROVED** [review]
Thanks for the contribution! I haven't tested it as I don't have OpenVino.

**Comment by dkurt** [comment]
Thanks for review! I will help with maintenance of that part so feel free to ping me in case of any issue :smile:

**Comment by tankienleong** [comment]
Screenshot from 2022-11-14 11-46-28

Ubuntu 20.04
deeplabcut 2.3rc2


Hi @dkurt , do you have any suggestion to solve this issue?

**Comment by dkurt** [comment]
@tankienleong, hi! Can you install with no dependencies?
```
pip install --no-deps openvino-dev
pip install openvino
```

**Comment by tankienleong** [comment]
Screenshot from 2022-11-15 09-52-57

@dkurt 
Do we need to solve those dependency conflicts?

Second question: do I need to remove RTX3060 from motherboard if I want to test the openvino?

**Comment by dkurt** [comment]
@tankienleong, everything seems good. No need remove GPU - OpenVINO will work as expected.

**Comment by mczhao1977** [comment]
@dkurt  Does pytorch backend of DLC 3.0 support OpenVINO? Thanks
